addons/multi_company_guidelines/models/models.py

```diff
@@ -0,0 +1,298 @@
+            if shared_record_companies and shared_record_companies != record.company_id:
+                errors.append(
+                    _(
+                        "\n - Some master data records belong to a different company than the one from the document."
+                    )
+                )
+                # here you could extend the error message to specify the name of records that are not shared and not in
+                # the same company to help the user correct the issue more easily
+                # do this carefully since it might degrade perfs or might be inconsistent in the UI
+                # if you use SQL to fetch non-translated values of translatable fields
+            if errors:
+                base_msg = _(
+                    'There is an inconsistency in the document "%s" which would cause multi-company issues'
+                    % (record.name,)
+                )
+                show_details = self.env.user.has_group(
+                    "base.group_user"
+                )  # don't leak data to non-internal users
+                details_msg = errors if show_details else ""
+                print("".join([base_msg] + details_msg))
+                raise ValidationError("".join([base_msg] + details_msg))
+
+    def _prepare_next_values(self):
+        """
+        Prepare a dictionnary of values for a new record in the flow.
+        
+        You should ensure that force_company is in the context with the company of
+        the parent record (for company dependent fields) and that company_id fields (or the
+        relationnal field from which the company comes in the case of a stored related field)
+        is present in the dictionnary of values!
+        """
+        self.ensure_one()
+        return {
+            "name": "Next from main record %s" % self.id,
+            "company_id": self.company_id.id,
+        }
+
+    def create_next_record(self):
+        record = self.with_context(force_company=self.company_id.id)
+        vals = record._prepare_next_values()
+        return record.env["multi_company.next"].create(vals)
+
+    def _prepare_post_values(self):
+        self.ensure_one()
+        # get the 'source' object; typically a bit more complex. Think getting
+        # the default sales journal of the company...
+        magic_source_record = self.env["multi_company.source"].search(
+            [("company_id", "=", self.company_id.id)], limit=1
+        )
+        if not magic_source_record:
+            raise UserError(_("There is no source model."))
+        return {"source_id": magic_source_record.id}
+
+    def create_post_record(self):
+        record = self.with_context(force_company=self.company_id.id)
+        vals = record._prepare_post_values()
+        return record.env["multi_company.related"].create(vals)
+
+    def action_confirm(self):
+        self.company_consistence_check()
+        for record in self:
+            record.create_next_record()
```

Comment:
making this kind of method (create_next_record) multi could lead to important performance improvements

here, if we want to make it multi and still have a correct force_company later on, we have to group self by company_id and call create_next_record group by group (we did that in _run_pull in master)

Reply:
reduce readability but you're right